lib/assert.js

```diff
@@ -608,6 +613,152 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
   }
 };
 
+/**
+ * Compares two objects or values recursively to check if they are equal.
+ * @param {any} actual - The actual value to compare.
+ * @param {any} expected - The expected value to compare.
+ * @param {boolean} [loose=false] - Whether to use loose comparison (==) or strict comparison (===). Defaults to false.
+ * @param {Set} [comparedObjects=new Set()] - Set to track compared objects for handling circular references.
+ * @returns {boolean} - Returns `true` if the actual value matches the expected value, otherwise `false`.
+ * @example
+ * // Loose comparison (default)
+ * compareBranch({a: 1, b: 2}, {a: 1, b: '2'}); // true
+ *
+ * // Strict comparison
+ * compareBranch({a: 1, b: 2}, {a: 1, b: 2}, true); // true
+ */
+function compareBranch(
+  actual,
+  expected,
+  loose = false,
+  comparedObjects = new Set(),
+) {
+  // Check for Date object equality
+  if (actual instanceof Date && expected instanceof Date) {
+    return actual.getTime() === expected.getTime();
+  }
+  // Check for RegExp object equality
+  else if (actual instanceof RegExp && expected instanceof RegExp) {
+    return actual.toString() === expected.toString();
+  } 
+  // Check non object types equality
+  else if (typeof actual !== 'object' || typeof expected !== 'object') {
```

Comment:
The `else` is redundant, as the previous `if`s return, and this is inconsistent with the rest of the function
```suggestion
  if (actual instanceof RegExp && expected instanceof RegExp) {
    return actual.toString() === expected.toString();
  } 
  // Check non object types equality
  if (typeof actual !== 'object' || typeof expected !== 'object') {
```